### PR TITLE
Resturcture of console and logic to allow for future expandability

### PR DESCRIPTION
**Pull request by Byron-Miles**
I've updated console.lua to parse the /buffduty command using the new more flexable logic and adjusted core.lua to match.

Note that this is currently un-tested and HAS BROKEN caching and assignment logic, which need to updated as well; however, I want to seperate the raid scanning logic out first.

I've also added a localisation.lua and moved the supported classes and channels into it (as well as re-working them a bit) in a step towards localisation - though I'm not really sure how that's all going to work at this point.

### CODE REVIEW

**Comment by Byron-Miles** [comment]
Seperated raid scanning logic and overhauled assignment logic. This is still un-tested and caching IS BROKEN.

TODO:
- Update caching to use raid-info tables
- Test and fix bugs
- Implement direct assign logic

**Comment by Byron-Miles** [comment]
I've reworked the Cache to use the new raid info tables, and added format versioning and a timeout to (hopefully) avoid format errors and the build up too much old data. See what you think.

I've also added the assignment logic, which should be the last of the structural overhauls (at least for now)... now to test and fix all the explody bits.

**Comment by Byron-Miles** [comment]
Done some testing, including on the new cache implementation, fixed some bugs and things seem to be working now :)

File: cache.lua [hunk 1]
```diff
@@ -1,27 +1,86 @@
...
+local utils = BuffDuty.Utils
+
+local function getTime()
+    return GetServerTime() -- WOW API: https://wow.gamepedia.com/API_GetServerTime
+end
+
+function Cache.generateHash(raid_info, class_players)
+    -- Raid groups
+    local raid_groups = {}
+    for i = 1, 8 do
```
Comment (azurefireice) [inline cache.lua:19]: what if raid has less then 8 groups?

File: localisation.lua [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+local SUPPORTED_CLASSES = { 
+    -- English
+    ["MAGE"] = CLASSES.MAGE,
+    ["M"] = CLASSES.MAGE,
+    ["PRIEST"] = CLASSES.PRIEST,
+    ["P"] = CLASSES.PRIEST,
+    ["DRUID"] = CLASSES.DRUID,
+    ["D"] = CLASSES.DRUID,
+    ["PALADIN"] = CLASSES.PALADIN,
+    ["PA"] = CLASSES.PALADIN,
```
Comment (azurefireice) [inline localisation.lua]: let's add all the rest classes here.

Comment (azurefireice) [inline localisation.lua]:
```suggestion
    ["PA"] = CLASSES.PALADIN,
    ["HUNTER"] = CLASSES.HUNTER,
    ["H"] = CLASSES.HUNTER,
    ["ROGUE"] = CLASSES.ROGUE,
    ["R"] = CLASSES.ROGUE,
    ["SHAMAN"] = CLASSES.SHAMAN,
    ["S"] = CLASSES.SHAMAN,
    ["WARLOCK"] = CLASSES.WARLOCK,
    ["W"] = CLASSES.WARLOCK,
    ["WARRIOR"] = CLASSES.WARRIOR,
    ["WARR"] = CLASSES.WARRIOR,
```

File: localisation.lua [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+BuffDuty.SUPPORTED_CHANNELS = SUPPORTED_CHANNELS
+
+-- CLASSES
+local CLASSES = {}
+-- English
+CLASSES.MAGE = "MAGE"
+CLASSES.PRIEST = "PRIEST"
+CLASSES.DRUID = "DRUID"
+CLASSES.PALADIN = "PALADIN"
```
Comment (azurefireice) [inline localisation.lua:30]:
```suggestion
CLASSES.PALADIN = "PALADIN"
CLASSES.HUNTER= "HUNTER"
CLASSES.ROGUE= "ROGUE"
CLASSES.SHAMAN= "SHAMAN"
CLASSES.WARLOCK= "WARLOCK"
CLASSES.WARRIOR= "WARRIOR"
```

File: raidinfo.lua [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    -- WOW API: https://wowwiki.fandom.com/wiki/API_GetRaidRosterInfo
+    local name, rank, sub_group, level, class_loc, class = GetRaidRosterInfo(idx)
+    return name, class, sub_group
+end
+
+local function getRaidMemberCount()
+    return GetNumGroupMembers() -- WOW API: https://wowwiki.fandom.com/wiki/API_GetNumGroupMembers
+end
+
+-- Scans the raid returns the following two data tables:
```
Comment (azurefireice) [inline raidinfo.lua]:
```suggestion
-- Scans the raid returns two data tables `raid_info`, `class_players`.
-- Table structure:
```

File: utils.lua [hunk 5]
```diff
@@ -1,12 +1,17 @@
 local Utils = {}
 BuffDuty.Utils = Utils
 
-function Utils.tableContainsValue(table, val)
-    if not table or not val then
+
+function Utils.containsName(table, name, get_name_func)
```
Comment (azurefireice) [inline utils.lua:17]: I do not understand these arguments. While `table, name,` make sense, i do not see the necessity of `get_name_func`.
Also, it is not used in code of `raidinfo`: 
image

File: utils.lua [hunk 6]
```diff
@@ -45,12 +50,31 @@ end
...
+    local idx = 0
     for value in string.gmatch(input, "([^"..seperator.."]+)") do -- Match all characters between seperators
-        table.insert(list, value)
+        idx = idx + 1
+        list[idx] = value
     end
     return list
 end
 
+function Utils.stringSplitAsFlags(input, seperator)
```
Comment (azurefireice) [inline utils.lua]:
```suggestion
function Utils.stringSplitAsFlags(input, separator)
```

File: utils.lua [hunk 7]
```diff
@@ -45,12 +50,31 @@ end
...
+    local list = {}
+    for value in string.gmatch(input, "([^"..seperator.."]+)") do -- Match all characters between seperators
```
Comment (azurefireice) [inline utils.lua]:
```suggestion
    for value in string.gmatch(input, "([^"..separator.."]+)") do -- Match all characters between separators
```

File: core.lua [hunk 8]
```diff
@@ -13,56 +17,75 @@ function BuffDuty:OnInitialize()
...
+        single_whisper = nil,
+        -- Flags
+        cache = true,
+        debug = false,
+    }
 
-    local cmd = {}
-    cmd.class = class
-    cmd.channel_name = channel_name
+    if BuffDuty.Console.parseDutyCommand(cmd, LibStub("AceConsole-3.0"):GetArgs(input, 20)) then
```
Comment (azurefireice) [inline core.lua]: Probably would be better to move `LibStub("AceConsole-3.0"):GetArgs(input, 20)` to a separate var.

File: core.lua [hunk 9]
```diff
@@ -82,7 +105,7 @@ local function executeMessage(input)
         verbose = false
     }
 
-    if(BuffDuty.Console.parseMessageCommand(cmd, LibStub("AceConsole-3.0"):GetArgs(input, 12))) then
+    if(BuffDuty.Console.parseMessageCommand(cmd, LibStub("AceConsole-3.0"):GetArgs(input, 14))) then
```
Comment (azurefireice) [inline core.lua]: same here

File: logic.lua [hunk 10]
```diff
@@ -1,167 +1,406 @@
...
-    local name, r, sg, lvl, cls_loc, cls = GetRaidRosterInfo(idx)
-    if sg > BuffDuty.max_group then
-        BuffDuty.max_group = sg
+-- Generates and returns a duty list in the following format:
+-- list[name]["name"] = the players name
+-- list[name]["i"] = the players index truncated to between 1 and 8 (inclusive)
+-- list[name]["groups"] = a string list of assigned groups
+-- list[name]["s"] = plural modifier; "s" if the player is assigned more than 1 group
+function BuffDuty.generateDuties(cmd, raid_info, class_players)
+    -- No players to assign :(
```
Comment (azurefireice) [inline logic.lua:13]: I really like smileys in comments 😃

File: logic.lua [hunk 11]
```diff
@@ -1,167 +1,406 @@
...
+            duty["s"] = "s"
+            duty["groups"] = string.format("%d - %d", raid_info.group_min, raid_info.group_max)
+        end
+        local duty_list = {}
+        duty_list[name] = duty
+        return duty_list
+    end
+
+    local player_duty_map = nil
+    if cmd.assign then
```
Comment (azurefireice) [inline logic.lua]: what are the options? Except assign?

File: logic.lua [hunk 12]
```diff
@@ -1,167 +1,406 @@
---local MAX_GROUPS = 8
```
Comment (azurefireice) [inline logic.lua:1]: @Byron-Miles What do you say if we move assignment strategies to a separate files?

**Comment by azurefireice** [comment]
Generally I understood everything and it makes sense to me. The assignment logic in second case seems unnecessarily complicated, but if you believe we need it - let's leave it, just let's store them in separate files. 400-line file is no good. 
There are also minor fixes and questions.
Overall - very good!

File: localisation.lua (hunk 2 above)
Comment (Byron-Miles) [reply localisation.lua]: I'm wondering if we want to depricate the short class names... it works OK for most classes, but Warrior and Paladin are a bit, hmm. And it's not like it saves that many characters.

File: utils.lua (hunk 5 above)
Comment (Byron-Miles) [reply utils.lua:17]: get_name_func allows you to pass in a custom function for getting the 'name' from the table... for example: `containsName(map, name, function(x) return x.name end)` where map has the structure `map[key].name `... there is only one place it is used like though, so could just make a seperate function.

File: core.lua (hunk 8 above)
Comment (Byron-Miles) [reply core.lua]: The issue is `GetArgs( )` returns multiple values.. putting it in a seperate var means using `pack()` and `unpack()` to pass it nicely to `parseCommand`... having it as an argument to the function call means it goes directly into the `...` parameter which is then easily convertable to a table using `{...}`

File: logic.lua (hunk 11 above)
Comment (Byron-Miles) [reply logic.lua]: Default? ... I was going to add more, but really Assign and Default - with the option to disable own group assignment - is pretty much it; unless you have some different assignment logic you want to use?

File: logic.lua (hunk 12 above)
Comment (Byron-Miles) [reply logic.lua:1]: That was my original plan... but then I couldn't really think of anything outside of the two existing logic types and it was just easier to lump them in the same file.

File: localisation.lua (hunk 2 above)
Comment (azurefireice) [reply localisation.lua]: Well, that would be a breaking change. If it's not too much trouble let's leave it until we implement UI.

File: utils.lua (hunk 5 above)
Comment (azurefireice) [reply utils.lua:17]: Yes, let's do it separate, since it's only 1 place where it's used.

File: core.lua (hunk 8 above)
Comment (azurefireice) [reply core.lua]: I mean like that. Are you sure we will need to pack/unpack in such case?
```suggestion
    local args = LibStub("AceConsole-3.0"):GetArgs(input, 20)
    if BuffDuty.Console.parseDutyCommand(cmd, args) then
```

File: core.lua (hunk 9 above)
Comment (azurefireice) [reply core.lua]: I can propose the following format:
```suggestion
    local args = LibStub("AceConsole-3.0"):GetArgs(input, 14)
    if(BuffDuty.Console.parseMessageCommand(cmd, args)) then
```

Also what do you think about having a function like `validateMessageCommand`? Code would look like 
```suggestion
    if(BuffDuty.Console.isCommandValid(args)) then
        cmd = BuffDuty.Console.parseMessageCommand(args)
```

File: logic.lua (hunk 12 above)
Comment (azurefireice) [reply logic.lua:1]: Let's do that anyway, to decouple strategy from controller. Once again, having a file with 400 lines is too much.

**Comment by azurefireice** [comment]
@Byron-Miles Please update the readme.md and we are good to merge. 
Also, do you think we need to cover some of this functionality with tests?

File: core.lua (hunk 8 above)
Comment (Byron-Miles) [reply core.lua]: Yes. 'args' in your example will only capture the first returned value (the rest are discarded)... you would need to use at least `local args = pack(LibStub("AceConsole-3.0"):GetArgs(input, 20))`

File: core.lua (hunk 9 above)
Comment (Byron-Miles) [reply core.lua]: It's kind of redundate, as you more or less have to parse the command to validate it... it also makes things more complex and prone to error as now you have to synchronise `isValid` and `parse` to ensure they are checking and reading the same values; much simpler and easier to do it all in one go unless there is a good reason you need to call `isValid` in multiple places and `parse` has an unacceptable performance overhead.

**Comment by Byron-Miles** [comment]
To the 2nd part, a definite yes. I have only had time to do basic tests so far and don't feel comfortable that sufficient coverage has been achieved to call it 'ready'.

File: core.lua (hunk 8 above)
Comment (Byron-Miles) [reply core.lua]: Also turns out `pack` (or rather `table.pack`) wasn't introduced until Lua 5.2 ... I've defined a local function instead.

File: core.lua (hunk 9 above)
Comment (azurefireice) [reply core.lua]: Then, parseMessageCommand should raise an error if command is invalid. It's very counter-intuitive that function named "parseMessageCommand" has to return boolean value, not the parsed command.

**Review by azurefireice — APPROVED** [review]
For the most part it is ready for merging. Please feel free to merge once you are confident. I'll update some parts like readme and will prepare a new release.

**Pull request merged**